src/components/pages/admin/contacts/RoleEditor.vue

```diff
@@ -0,0 +1,167 @@
+<template>
+  <div>
+    <div
+      v-for="role in contact.roles"
+      :key="role.role + role.dateAdded.toJSON()"
+    >
+      <RoleEditorItem
+        :key="role.role + role.dateAdded.toJSON()"
+        :role="role"
+        :contact="contact"
+        @update="emit('update')"
+      />
+    </div>
+
+    <div class="mt-3">
+      <AppButton
+        class="w-full !bg-primary-5"
+        variant="primaryOutlined"
+        @click="formVisible = true"
+        >{{ t('contacts.data.roles.add') }}</AppButton
+      >
+    </div>
+
+    <div
+      v-if="formVisible"
+      class="flex flex-1 rounded border border-primary-20 bg-primary-10 p-4"
+    >
+      <form class="flex-initial" @submit.prevent="handleFormSubmit">
+        <div class="my-2 py-1">
+          <AppSelect
+            v-model="newRole.role"
+            label="New role"
+            :items="roleOptions"
+          />
+        </div>
+
+        <div>
+          <div class="my-2 py-1">
+            <AppLabel :label="inputT('starts.label')" required />
+            <AppRadioGroup
+              v-model="roleHasStartDate"
+              name="roleStartDate"
+              :options="[
+                [false, inputT('starts.opts.now')],
+                [true, inputT('starts.opts.schedule')],
+              ]"
+              required
+            />
+            <div v-if="roleHasStartDate" class="flex gap-2">
+              <div>
+                <AppInput v-model="newRole.startDate" type="date" required />
+              </div>
+              <div>
+                <AppInput v-model="newRole.startTime" type="time" required />
+              </div>
+            </div>
+          </div>
+
+          <div class="my-2 py-1">
+            <AppRadioGroup
+              v-model="roleHasEndDate"
+              name="roleEndDate"
+              :label="inputT('expires.label')"
+              :options="[
+                [false, inputT('expires.opts.never')],
+                [true, inputT('expires.opts.schedule')],
+              ]"
+              required
+            />
+            <div v-if="roleHasEndDate" class="flex gap-2">
+              <div>
+                <AppInput v-model="newRole.endDate" type="date" required />
+              </div>
+              <div>
+                <AppInput v-model="newRole.endTime" type="time" required />
+              </div>
+            </div>
+          </div>
+        </div>
+
+        <div class="my-3 flex py-3">
+          <AppButton type="submit" variant="primary" :loading="loading">{{
```

Comment:
The button should not be enabled unless the form validates. In practice this is very easy to achieve as long as all the validation props (`required`, etc.) are set correctly which they appear to be here.

```vue
<AppButton ... :disabled="validation.$invalid" />
```

In the script
```ts
const validation = useVuelidate();
```